assets/pages/new-contributions-landing/components/SetPassword/SetPassword.jsx

```diff
@@ -0,0 +1,154 @@
+
+import { type Contrib } from 'helpers/contributions';
+import { classNameWithModifiers } from 'helpers/utilities';
+
+import SvgPasswordKey from 'components/svgs/passwordKey';
+import SvgEnvelope from 'components/svgs/envelope';
+import CtaLink from 'components/ctaLink/ctaLink';
+import type { Csrf as CsrfState } from 'helpers/csrf/csrfReducer';
+import { NewContributionTextInput } from '../ContributionTextInput';
+import { CreateAccountButton } from './CreateAccountButton';
+import { type ThankYouPageStage } from '../../contributionsLandingReducer';
+import { setThankYouPageStage, setPasswordHasBeenSubmitted, updatePassword, type Action } from '../../contributionsLandingActions';
+import { setPasswordGuest } from './helpers';
+import { checkEmail, emailRegexPattern } from '../../formValidation';
+
+// ----- Types ----- //
+
+/* eslint-disable react/no-unused-prop-types */
+type PropTypes = {
+  contributionType: Contrib,
+  email: string,
+  password: string,
+  guestAccountCreationToken: string,
+  setThankYouPageStage: (ThankYouPageStage) => void,
+  setPasswordHasBeenSubmitted: () => void,
+  passwordHasBeenSubmitted: boolean,
+  updatePassword: (Event) => void,
+  csrf: CsrfState,
+};
+/* eslint-enable react/no-unused-prop-types */
+
+
+// ----- State Maps ----- //
+
+const mapStateToProps = state => ({
+  contributionType: state.page.form.contributionType,
+  email: state.page.form.formData.email,
+  password: state.page.form.setPasswordData.password,
+  passwordHasBeenSubmitted: state.page.form.setPasswordData.passwordHasBeenSubmitted,
+  guestAccountCreationToken: state.page.form.guestAccountCreationToken,
+  csrf: state.page.csrf,
+});
+
+function mapDispatchToProps(dispatch: Dispatch<Action>) {
+  return {
+    setThankYouPageStage: (thankYouPageStage: ThankYouPageStage) => {
+      dispatch(setThankYouPageStage(thankYouPageStage));
+    },
+    setPasswordHasBeenSubmitted: () => {
+      dispatch(setPasswordHasBeenSubmitted());
+    },
+    updatePassword: (event: Event) => {
+      if (event.target instanceof HTMLInputElement) {
+        dispatch(updatePassword(event.target.value));
+      }
+    },
+  };
+}
+
+
+// ----- Functions ----- //
+
+function onSubmit(props: PropTypes): Event => void {
+  return (event) => {
+    props.setPasswordHasBeenSubmitted();
+    event.preventDefault();
+
+    if (!(event.target: any).checkValidity()) {
+      return;
+    }
+
+    // TODO: send user to thank you page with error if password was not set
+    setPasswordGuest(props.password, props.guestAccountCreationToken, props.csrf)
+      .then((response) => {
+        if (response === true) {
+          props.setThankYouPageStage('thankYouPasswordSet');
+        } else {
+          props.setThankYouPageStage('thankYouPasswordNotSet');
+        }
+      });
+  };
+}
+
+
+// ----- Render ----- //
+
+function SetPassword(props: PropTypes) {
+  if (!props.guestAccountCreationToken || !props.email) {
+    props.setThankYouPageStage('thankYou');
```

Comment:
It looks like you could move that logic up the call chain, right in the reducer and save the browser some work.

Reply:
like this?

```
      // Don't allow the stage to be set to setPassword unless both an email and
      // guest registration token is present
      case 'SET_THANK_YOU_PAGE_STAGE':
        if ((action.thankYouPageStage === 'setPassword')
          && (!state.guestAccountCreationToken || !state.formData.email)) {
          return { ...state, thankYouPageStage: 'thankYou' };
        }
        return { ...state, thankYouPageStage: action.thankYouPageStage };
```

Reply:
yes